### PR TITLE
Taking on every NYI we hit while parsing a large code base

### PR DESCRIPTION
**Pull request by jurgenvinju**
This is to fix this stracktrace:

image

### CODE REVIEW

**Comment by jurgenvinju** [comment]
This needs some testing. @linuswagner could you give it as spin? and perhaps supply a small code example that exercises this new feature?

File: src/lang/cpp/AST.rsc [hunk 1]
```diff
@@ -554,3 +554,59 @@ java rel[loc,loc] parseForMacros(loc file, str charset=DEFAULT_CHARSET, bool inf
 @synopsis{All functions in this module that have a charset parameter use this as default.}
 public str DEFAULT_CHARSET = "UTF-8";
 
+public map[str, str] linusMacros = (
```
[the remark this answers is not in the document]
Comment (jurgenvinju) [reply src/lang/cpp/AST.rsc:557]: strange.. I must have `git commit -am`ed again.. it's in my muscle memory.

[the remark this answers is not in the document]
Comment (jurgenvinju) [reply src/lang/cpp/AST.rsc]: What's the status now? are we ready to merge?

File: src/lang/cpp/TypeSymbol.rsc [hunk 2]
```diff
@@ -75,11 +75,12 @@ data TypeSymbol
   | \functionTemplate(loc decl, list[loc] templateParameters)
   | \variableTemplate(loc decl, list[loc] templateParameters)
   
-  | \aliasTemplate(loc decl, list[loc] templateParameters)
+  | \aliasTemplate(loc decl, list[loc] templateParameters, TypeSymbol \type)
   
   | \functionSetType(loc decl, list[TypeSymbol] templateArguments)
   | \functionSetTypePointer(loc decl, list[TypeSymbol] templateArguments)
   
+  | \unaryTypeTransformation(str operator, TypeSymbol operand)
```
Comment (rodinaarssen) [inline src/lang/cpp/TypeSymbol.rsc:83]: I'm not a big fan of encoding the operator as a string parameter. Can we specialize this per operator? Looking at the code, there's (currently) only one value in the operator enum anyway

File: src/lang/cpp/internal/BindingsResolver.java [hunk 3]
```diff
@@ -376,7 +376,7 @@ private ISourceLocation resolveIEnumeration(IEnumeration binding, ISourceLocatio
 		throw new RuntimeException("NYI" + binding.getClass().getSimpleName() + ": " + binding + " @ " + origin);
 	}
 
-	private ISourceLocation resolveICPPBinding(ICPPBinding binding, ISourceLocation origin) throws URISyntaxException {
```
Comment (rodinaarssen) [inline src/lang/cpp/internal/BindingsResolver.java:379]: Was this change intentional?

File: src/lang/cpp/internal/BindingsResolver.java [hunk 4]
```diff
@@ -840,6 +840,16 @@ public ISourceLocation resolveBinding(IASTNameOwner node, ISourceLocation origin
 		return failedBinding("unknown");
 	}
 
+	public ISourceLocation resolveBinding(ICPPBinding binding, ISourceLocation origin) {
```
Comment (rodinaarssen) [inline src/lang/cpp/internal/BindingsResolver.java:843]: Is there a particular reason that the preexisting interface method did not suffice?

File: src/lang/cpp/internal/Parser.java [hunk 5]
```diff
@@ -1459,29 +1457,6 @@ private int visit(ICPPASTDesignatedInitializer initializer) {
 		return PROCESS_ABORT;
 	}
 
-	// InitializerClauses
```
Comment (rodinaarssen) [inline src/lang/cpp/internal/Parser.java:1462]: Was this unreachable code?

File: src/lang/cpp/internal/Parser.java [hunk 6]
```diff
@@ -1557,12 +1532,6 @@ private int visit(IASTArrayDeclarator declarator) {
 		return PROCESS_ABORT;
 	}
 
-	private int visit(IASTFieldDeclarator declarator) {
```
Comment (rodinaarssen) [inline src/lang/cpp/internal/Parser.java:1560]: Was this unreachable code?

**Review by rodinaarssen — COMMENTED** [review]
Left some comments here and there. Looks good in general, but this PR has become much larger than just the alias templates, which makes it harder to review.

File: src/lang/cpp/TypeSymbol.rsc (hunk 2 above)
Comment (jurgenvinju) [reply src/lang/cpp/TypeSymbol.rsc:83]: I don't like that either. I'll have a look. If it's a small finite set we can have one transformation constructor per operator. Otherwise, we introduce an TypeTransformationOperator sort.

File: src/lang/cpp/internal/BindingsResolver.java (hunk 3 above)
Comment (jurgenvinju) [reply src/lang/cpp/internal/BindingsResolver.java:379]: Yes this is dead code.

File: src/lang/cpp/internal/BindingsResolver.java (hunk 4 above)
Comment (jurgenvinju) [reply src/lang/cpp/internal/BindingsResolver.java:843]: Yes, due to multiple inheritance of interfaces, the code for reaching ICPPBinding was not always reachable, in particular when the object did not accidentally also implement IASTNameOwner.

File: src/lang/cpp/internal/Parser.java (hunk 5 above)
Comment (jurgenvinju) [reply src/lang/cpp/internal/Parser.java:1462]: yes. the compiler warned about that

**Comment by jurgenvinju** [comment]
> Left some comments here and there. Looks good in general, but this PR has become much larger than just the alias templates, which makes it harder to review.

So true. We are under a lot of time pressure here so we decided to take a sprint and solve every remaining issue as it hits us. The code base we are trying to parse and map is incredibly large and decades old, so it seems we are bound for hitting every `NYI` on the way.

**Comment by jurgenvinju** [comment]
Just added GNU's computed goto's. Will try and simplify the TypeTransformationOperator design now.
